assembly/src/parsers/ast/mod.rs

```diff
@@ -0,0 +1,563 @@
+        &mut self,
+        tokens: &mut TokenStream,
+        program: &mut types::Program,
+    ) -> Result<(), AssemblyError> {
+        while let Some(token) = tokens.read() {
+            match token.parts()[0] {
+                Token::EXPORT => {
+                    let procedure = self.parse_procedure(tokens, program, true)?;
+                    let name = procedure.name.to_string();
+                    program.procedures.insert(name, procedure);
+                }
+                Token::PROC => {
+                    let mut procedure = self.parse_procedure(tokens, program, false)?;
+                    procedure.index = program.procedures.len() as u32;
+                    let name = procedure.name.to_string();
+                    program.procedures.insert(name, procedure);
+                }
+                _ => break,
+            }
+        }
+
+        Ok(())
+    }
+
+    // Parses the token streams into AST nodes
+    fn parse_body(
+        &mut self,
+        tokens: &mut TokenStream,
+        program: &mut types::Program,
+    ) -> Result<(), AssemblyError> {
+        let program_start = tokens.pos();
+        // consume the 'begin' token
+        let header = tokens.read().expect("missing program header");
+        header.validate_begin()?;
+        tokens.advance();
+
+        // make sure there is something to be read
+        let start_pos = tokens.pos();
+        if tokens.eof() {
+            return Err(AssemblyError::unexpected_eof(start_pos));
+        }
+
+        // parse the sequence of nodes and add each node to the list
+        let mut end_of_nodes = false;
+        let mut empty = true;
+        while !end_of_nodes {
+            let parsed_nodes = self.parse_body_token(tokens, program)?;
+            end_of_nodes = parsed_nodes.is_empty();
+            for node in parsed_nodes {
+                program.body.push(node);
+                empty = false;
+            }
+        }
+
+        // make sure at least one block has been read
+        if empty {
+            let start_op = tokens.read_at(start_pos).expect("no start token");
+            return Err(AssemblyError::empty_block(start_op));
+        }
+
+        // consume the 'end' token
+        match tokens.read() {
+            None => Err(AssemblyError::unmatched_begin(
+                tokens.read_at(program_start).expect("no begin token"),
+            )),
+            Some(token) => match token.parts()[0] {
+                Token::END => token.validate_end(),
+                Token::ELSE => Err(AssemblyError::dangling_else(token)),
+                _ => Err(AssemblyError::unmatched_begin(
+                    tokens.read_at(program_start).expect("no begin token"),
+                )),
+            },
+        }?;
+        tokens.advance();
+
+        Ok(())
+    }
+
+    // Parse procedure from token stream.
+    fn parse_procedure(
+        &mut self,
+        tokens: &mut TokenStream,
+        program: &mut types::Program,
```

Comment:
Why does `program` need to be mutable here? Also, it would probably be better to pass just a map of procedures rather than a full `Program` struct into this. This way, we could reuse this function for other purposes (e.g., parsing library modules).

Reply:
fixed